Feat footnote

Jocs opened the pull request:
| Q                | A
| ---------------- | ---
| New feature?     | yes
| Fixed tickets    | #193 #1568 
| License          | MIT

### Description

<img width="802" alt="屏幕快照 2019-10-27 下午7 06 05" src="https://user-images.githubusercontent.com/9712830/67633614-f4877880-f8ec-11e9-8c3b-d54a1baf6c1a.png">

- [x] Create footnote from `footnote identifier`
- [x] Jump to footnote from `footnote identifier`
- [x] Jump to `footnote identifier` from `footnote`
- [x] Handle `backspace` in footnote
- [x] Export markdown with footnote
- [x] Import markdown with footnote
- [x] Footnote setting option in setting window

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/marktext/marktext/1546)
<!-- Reviewable:end -->

fxha commented on src/muya/lib/parser/index.js:
```diff
@@ -223,6 +223,27 @@ const tokenizerFac = (src, beginRules, inlineRules, pos = 0, top, labels) => {
         continue
       }
     }
+    // footnote identifier
```
Newline is recommended.

fxha commented on src/muya/lib/parser/marked/utils.js:
```diff
@@ -2,6 +2,10 @@
  * Helpers
  */
 
+let i = 0
```
Please use a different variable name like `uniqueIdCounter`.

Jocs commented on src/muya/lib/parser/marked/lexer.js:
```diff
@@ -28,15 +29,39 @@ Lexer.prototype.lex = function (src) {
   this.checkFrontmatter = true
-  return this.token(src, true)
+  this.token(src, true)
+  // Move footnote token to the end of tokens.
```
Export html/PDF also need to set footnotes at the end of document.

fxha replied:
That's why add an option to allow to disable it.

Jocs commented:
>Table inside footnote block doesn't work - no design and seems buggy.

It seems work.
![屏幕快照 2019-10-30 下午12 45 30](https://user-images.githubusercontent.com/9712830/67829749-3c87e480-fb13-11e9-8bb4-4c31eefad656.png)

Jocs commented:
>Cannot create paragraph behind a code/HTML/math/diagram block in footnote block and context menu create paragraph after just insert a new paragraph behind the footnote block.

Any ideas to create paragraph bellow these blocks? we can only create new paragraph bellow these blocks in source code mode. BTW, maybe not much users use multiple blocks footnote :smile:

Jocs commented:
>Please remove debug code

Fixed

>[^longnote] is not auto numbered in exported document but in footnotes.

Fixed

>Spell checking is enabled on ag-footnote-input (footnote identifer in display block)

Fixed

>Footnote display box should be also creatable with Enter (--> [^foo]:<enter>)

Fixed

>Table inside footnote block doesn't work - no design and seems buggy.

Need discuss.

>Cannot create paragraph behind a code/HTML/math/diagram block in footnote block and context menu create paragraph after just insert a new paragraph behind the footnote block.

Need discuss.

fxha commented:
Related block content: I think the use case is very limited but a footnote block is the same as a list or quote block, right? We could keep the issue with the new paragraph in mind but should support block elements inside.

Jocs commented:
>It would be cool to allow to click on footnote number in footnotes list in exported document to navigate to the footnote in text.

it seems that the number in front of list is not clickable...because it not an html element, but we can add a click element in future improvement.
